### PR TITLE
add kvstore server and client

### PR DESCRIPTION
**Pull request by hao-liu-1117**
Hi Professor,

Thanks for your reply. Maybe I've figure out how to build server and client properly with grpc and my current code.

I would be quick to finish other part.

Thanks.

### CODE REVIEW

File: src/kvstore_client.cc [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "kvstore_client.h"
+
+bool TestPrintPairs (const std::vector<std::string> &keyarr,
```
Comment (barath) [inline src/kvstore_client.cc]: Testing code should go in testing files.

File: src/kvstore_client.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  }
+  return true;
+}
+
+int main (int argc, char** argv) {
+  KVStoreClient client(grpc::CreateChannel("0.0.0.0:50001",
+                grpc::InsecureChannelCredentials()));
+  std::vector<std::string> keyarr, valarr;
+  for (int i = 0; i < 10; i++) {
+    keyarr.push_back("key"+std::to_string(i));
```
Comment (barath) [inline src/kvstore_client.cc]: Testing code should go in testing files only.

File: src/kvstore_client.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+#include "keyvaluestoreclass.h"
+#include "keyvaluestore.grpc.pb.h"
+
+// Key-value store client.
+class KVStoreClient {
+ public:
+  explicit KVStoreClient (std::shared_ptr<grpc::Channel> channel)
+                         : stub_(kvstore::KeyValueStore::NewStub(channel)) {}
+  
+  bool Put (const std::string &key, const std::string &value) {
```
Comment (barath) [inline src/kvstore_client.h]: Move full implementations to .cc files.

File: src/kvstore_client.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    grpc::Status status = stub_->Put(&context, request, &reply);
+
+    if (!status.ok()) {
+      std::cout << "status error: " << status.error_message() << std::endl;
+      return false;
+    }
+    return true;
+  }
+
+  bool Get (const std::vector<std::string> &keyarr, std::vector<std::string> &valarr) {
```
Comment (barath) [inline src/kvstore_client.h]: No space before (

File: src/kvstore_client.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  bool Remove(const std::string &key) {
+    grpc::ClientContext context;
+    kvstore::RemoveRequest request;
+    kvstore::RemoveReply reply;
+
+    request.set_key(key);
+    grpc::Status status = stub_->Remove(&context, request, &reply);
+
+    if (!status.ok()) {
+      std::cout << "status error: " << status.error_message() << std::endl;
```
Comment (barath) [inline src/kvstore_client.h]: Use LOG instead of cout

File: src/kvstore_client.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      return false;
+    }
+    return true;
+  }
+
+ private:
+  std::unique_ptr<kvstore::KeyValueStore::Stub> stub_;
```
Comment (barath) [inline src/kvstore_client.h:33]: Member var comment

File: src/kvstore_server.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <grpcpp/grpcpp.h>
+
+#include "keyvaluestoreclass.h"
+#include "keyvaluestore.grpc.pb.h"
+
+class KVStoreService final : public kvstore::KeyValueStore::Service {
+ public:
+  explicit KVStoreService () {}
+  grpc::Status Put(grpc::ServerContext* context, const kvstore::PutRequest* request,
```
Comment (barath) [inline src/kvstore_server.h:14]: Move impls to .cc files

File: src/kvstore_client.cc (hunk 1 above)
Comment (hao-liu-1117) [reply src/kvstore_client.cc]: I'll start test code in a new file.

File: src/kvstore_client.cc (hunk 2 above)
Comment (hao-liu-1117) [reply src/kvstore_client.cc]: Sure, I'll do that.

File: src/kvstore_client.h (hunk 3 above)
Comment (hao-liu-1117) [reply src/kvstore_client.h]: OK, I'll do that.

File: src/kvstore_client.h (hunk 4 above)
Comment (hao-liu-1117) [reply src/kvstore_client.h]: My bad, I'll change that.

File: src/kvstore_client.h (hunk 5 above)
Comment (hao-liu-1117) [reply src/kvstore_client.h]: Sure, I'll use LOG to track those info.

File: src/kvstore_client.h (hunk 6 above)
Comment (hao-liu-1117) [reply src/kvstore_client.h:33]: I'll add it on.

File: src/kvstore_server.h (hunk 7 above)
Comment (hao-liu-1117) [reply src/kvstore_server.h:14]: I'll do that.

**Review by hao-liu-1117 — COMMENTED** [review]
Currently I would put my server test code in /src because of some dependency problem, but I believe it is a minor problem.

I didn't find a way out so far but I'm sure that I could fix it out later.

Maybe I should move on to other components in quick for now.

**Comment by hao-liu-1117** [comment]
Hi Professor,

I've update my kvstore server and client code.

File: src/kvstore_client.cc [hunk 8]
```diff
@@ -1,45 +1,63 @@
...
+  kvstore::PutReply reply;
+
+  request.set_key(key);
+  request.set_value(value);
+  // Put function finished by server.
+  grpc::Status status = stub_->Put(&context, request, &reply);
+
+  if (!status.ok()) {
+    std::cout << "status error: " << status.error_message() << std::endl;
+  return false;
```
Comment (barath) [inline src/kvstore_client.cc]: Fix indent

File: src/kvstore_client.cc [hunk 9]
```diff
@@ -1,45 +1,63 @@
...
+  kvstore::PutRequest request;
+  kvstore::PutReply reply;
+
+  request.set_key(key);
+  request.set_value(value);
+  // Put function finished by server.
+  grpc::Status status = stub_->Put(&context, request, &reply);
+
+  if (!status.ok()) {
+    std::cout << "status error: " << status.error_message() << std::endl;
```
Comment (barath) [inline src/kvstore_client.cc]: Perhaps use LOG instead of cout

Comment (hao-liu-1117) [reply src/kvstore_client.cc]: I'll do that.

File: src/kvstore_client.cc (hunk 8 above)
Comment (hao-liu-1117) [reply src/kvstore_client.cc]: Sure.

**Review by barath — APPROVED** [review]

**Pull request merged**